### DISCUSSION TITLE
CPP17 branch example logging config

### DISCUSSION THREAD

**Discussion by tunguskar** [intel/compile-time-init-build#566] [Q&A] [answered]
I created this minimal example: 

```cpp
struct my_logger_config : logging::null::config
{
  struct
  {
    template <logging::level L, typename... Ts>
    constexpr auto log(Ts &&...args) const noexcept -> void
    {
      std::puts(std::forward<Ts>(args)...);
    }
  } logger;
  
  static auto terminate() noexcept -> void
  {
    //do something maybe reset?
  }
};

template <>
inline auto logging::config<> = my_logger_config{};

void testFunc(void)
{
  CIB_INFO("Hello World");
}
```

Has someone a clue how to call std::puts on this? Compiler says: 

```bash
too many arguments to function ‘int puts(const char*)’GCC
cib.hpp(26, 16): In instantiation of ‘constexpr void my_logger_config::<unnamed struct>::log(Ts&& ...) const [with logging::level L = logging::INFO; Ts = {const char (&)[73], int, sc::lazy_string_format<sc::string_constant<char, 'H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd'>, cib::tuple_impl<> >}]’:
```

**Comment by elbeno** [comment, marked as answer]
See how the fmt logger does it: https://github.com/intel/compile-time-init-build/blob/cpp17/include/log/fmt/logger.hpp#L29

That pack of Ts... contains several arguments:
`Ts = {const char (&)[73], int, sc::lazy_string_format<sc::string_constant<char, 'H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd'>, cib::tuple_impl<> >}`

- `const char (&)[73]` is `__FILE__`
- `int` is `__LINE__`
- `sc::lazy_string_format<sc::string_constant<char, 'H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd'>, cib::tuple_impl<> >` is the format string and any runtime arguments to put into it (in this case none)

`puts` isn't going to handle this in general (other than possibly outputting the `__FILE__` and the format string unaltered).

**Reply by tunguskar** [reply]
Just for explanation. I plan to optional abstract the hardware and run the system without the target. For this I need the Logging to the console. Thanks for the hint. In relation to this I found in your doc some alternative way:

```cpp
template <>
inline auto logging::config<> = logging::fmt::config{std::ostream_iterator<char>{std::cout}};
```
